torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp

```diff
@@ -176,12 +177,119 @@ struct CollectiveFingerPrint {
              << backend->getRank() << " is running collective: " << *this
              << ", but Rank " << rank
              << " is running collective: " << rank_fingerprint << ".";
+          auto diff_result = compute_collective_diff(rank_fingerprint);
+          auto has_diff = std::get<0>(diff_result);
+          if (std::get<0>(diff_result)) {
+            ss << std::get<1>(diff_result);
+          }
+
           TORCH_CHECK(false, ss.str());
         }
       }
     }
   }
 
+  static std::vector<std::string> get_size_strs(
+      const CollectiveFingerPrint& collective_fingerprint) {
+    std::vector<std::string> size_strs;
+    if (!collective_fingerprint.tensor_sizes_.empty()) {
+      for (const auto& single_tensor_shape_num :
+           collective_fingerprint.tensor_sizes_[0]) {
+        size_strs.emplace_back(std::to_string(single_tensor_shape_num));
+      }
+    }
+    return size_strs;
+  }
+
+  static std::vector<std::string> get_dtype_strs(
+      const CollectiveFingerPrint& collective_fingerprint) {
+    std::vector<std::string> dtype_strs;
+    dtype_strs.reserve(collective_fingerprint.tensor_dtypes_.size());
+    for (const auto& tensor_dtype : collective_fingerprint.tensor_dtypes_) {
+      dtype_strs.emplace_back(
+          c10::toString(static_cast<at::ScalarType>(tensor_dtype)));
+    }
+    return dtype_strs;
+  }
+
+  static std::vector<std::string> get_device_type_strs(
+      const CollectiveFingerPrint& collective_fingerprint) {
+    std::vector<std::string> device_type_strs;
+    device_type_strs.reserve(
+        collective_fingerprint.tensor_device_types_.size());
+    for (const auto& tensor_device_type :
+         collective_fingerprint.tensor_device_types_) {
+      device_type_strs.emplace_back(
+          c10::toString(static_cast<at::DeviceType>(tensor_device_type)));
+    }
+    return device_type_strs;
+  }
+
+  std::pair<bool, std::string> compute_collective_diff(
```

Comment:
We could save a lot of copying by passing the stringstream as argument, but it would be a huge seekg mess in case the diff is not found.

Reply:
Seems reasonable to use a stringstream, and rely on the bool return in the pair to indicate whether it is valid or not?